plugin/db/mongodb/migrate.go

```diff
@@ -85,16 +115,320 @@ func (driver *Driver) SetupMigrationIfNeeded(ctx context.Context) error {
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to aggregate migration history list to find largest version since baseline")
+	}
+	defer cursor.Close(ctx)
+
+	if cursor.Next(ctx) {
+		var result bson.M
+		if err := cursor.Decode(&result); err != nil {
+			return nil, errors.Wrapf(err, "failed to decode find largest version since baseline aggregation result")
+		}
+		if version, ok := result["version"]; ok {
+			if versionStr, ok := version.(string); ok {
+				return &versionStr, nil
+			}
+		}
+		return nil, errors.Errorf("failed to get version from find largest version since baseline aggregation result: %v", result)
+	}
+	// If cursor contains no data, we check the err, if err is nil, it means that the query is successful, but there is no data.
+	if err := cursor.Err(); err != nil {
+		return nil, errors.Wrapf(err, "failed to find largest version since baseline")
+	}
+	return nil, nil
+}
+
+// InsertPendingHistory will insert the migration record with pending status and return the inserted ID.
+func (driver *Driver) InsertPendingHistory(ctx context.Context, _ *sql.Tx, sequence int, _ string, m *db.MigrationInfo, storedVersion, statement string) (insertedID int64, err error) {
+	collection := driver.client.Database(migrationHistoryDefaultDatabase).Collection(migrationHistoryDefaultCollection)
+
+	retryTimes := 3
+	nextID := int64(0)
+	for i := 0; i < retryTimes; i++ {
+		currentTimestamp := getMongoTimestamp()
+		nextID, err = driver.getMigrationHistoryNextID(ctx)
+		if err != nil {
+			return 0, errors.Wrapf(err, "failed to get next migration history ID")
+		}
+
+		document := bson.M{
+			"id":                    nextID,
+			"created_by":            m.Creator,
+			"created_ts":            currentTimestamp,
+			"updated_by":            m.Creator,
+			"updated_ts":            currentTimestamp,
+			"release_version":       m.ReleaseVersion,
+			"namespace":             m.Namespace,
+			"sequence":              int64(sequence),
+			"source":                m.Source,
+			"type":                  m.Type,
+			"status":                db.Pending,
+			"version":               storedVersion,
+			"description":           m.Description,
+			"statement":             statement,
+			"execution_duration_ns": int64(0),
+			"issue_id":              m.IssueID,
+			"payload":               m.Payload,
+		}
+		if _, err = collection.InsertOne(ctx, document); err != nil {
+			// Detect if it's a duplicate key error.
+			if driverErr, ok := err.(mongo.WriteException); ok {
+				find := false
+				for _, writeErr := range driverErr.WriteErrors {
+					if writeErr.Code == 121 {
+						log.Info("Duplicate key error with migration_history id: %d, retrying...", zap.Int64("id", nextID))
+						find = true
+						break
+					}
+				}
+				if find {
+					continue
+				}
+			}
+			return 0, errors.Wrapf(err, "failed to insert a pending migration history record")
+		}
+		return nextID, nil
+	}
+	return nextID, nil
```

Comment:
This should return error since we have exhausted retry

Reply:
SGTM. Good catch.